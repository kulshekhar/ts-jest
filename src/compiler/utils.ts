import * as path from 'path';
import * as ts from 'typescript';
import * as fs from 'fs';

// from here: https://github.com/s-panferov/awesome-typescript-loader/blob/master/src/helpers.ts

export interface OutputFile {
  text: string;
  sourceMap?: string;
  declaration?: ts.OutputFile;
}

function statSyncNoException(path: string) {
  try {
    return fs.statSync(path);
  } catch (e) {
    return undefined;
  }
}

function withoutExt(fileName: string): string {
  return path.basename(fileName).split('.')[0];
}

let caseInsensitiveFs: boolean | undefined;

export function isCaseInsensitive() {
  if (typeof caseInsensitiveFs !== 'undefined') {
    return caseInsensitiveFs;
  }

  const lowerCaseStat = statSyncNoException(process.execPath.toLowerCase());
  const upperCaseStat = statSyncNoException(process.execPath.toUpperCase());

  if (lowerCaseStat && upperCaseStat) {
    caseInsensitiveFs =
      lowerCaseStat.dev === upperCaseStat.dev &&
      lowerCaseStat.ino === upperCaseStat.ino;
  } else {
    caseInsensitiveFs = false;
  }

  return caseInsensitiveFs;
}

function compareFileName(first: string, second: string) {
  if (isCaseInsensitive()) {
    return first.toLowerCase() === second.toLowerCase();
  } else {
    return first === second;
  }
}

function isFileEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    // typescript now emits .jsx files for .tsx files.
    (outputFileName.substr(-3).toLowerCase() === '.js' ||
      outputFileName.substr(-4).toLowerCase() === '.jsx')
  );
}

function isSourceMapEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    // typescript now emits .jsx files for .tsx files.
    (outputFileName.substr(-7).toLowerCase() === '.js.map' ||
      outputFileName.substr(-8).toLowerCase() === '.jsx.map')
  );
}

function isDeclarationEmit(
  fileName: string,
  outputFileName: string,
  sourceFileName: string,
) {
  return (
    compareFileName(sourceFileName, fileName) &&
    outputFileName.substr(-5).toLowerCase() === '.d.ts'
  );
}

export function findResultFor(
  fileName: string,
  output: ts.EmitOutput,
): OutputFile {
  let text: string | undefined;
  let sourceMap: string | undefined;
  let declaration: ts.OutputFile | undefined;
  fileName = withoutExt(fileName);

  for (let i = 0; i < output.outputFiles.length; i++) {
    let o = output.outputFiles[i];
    let outputFileName = o.name;
    let sourceFileName = withoutExt(o.name);
    if (isFileEmit(fileName, outputFileName, sourceFileName)) {
      text = o.text;
    }
    if (isSourceMapEmit(fileName, outputFileName, sourceFileName)) {
      sourceMap = o.text;
    }
    if (isDeclarationEmit(fileName, outputFileName, sourceFileName)) {
      declaration = o;
    }
  }
  if (text === undefined) {
    // console.log(text)
    throw new Error('text is undefined');
  }
  return {
    text,
    sourceMap,
    declaration,
  };
}
