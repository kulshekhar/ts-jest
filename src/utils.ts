import * as crypto from 'crypto';
import * as fs from 'fs';
import * as fsExtra from 'fs-extra';
import * as path from 'path';
import * as tsc from 'typescript';
import { ConfigGlobals, JestConfig, TsJestConfig } from './jest-types';
import { logOnce } from './logger';
import * as _ from 'lodash';

export function getTSJestConfig(globals: ConfigGlobals): TsJestConfig {
  return globals && globals['ts-jest'] ? globals['ts-jest'] : {};
}

function formatTsDiagnostics(errors: tsc.Diagnostic[]): string {
  const defaultFormatHost: tsc.FormatDiagnosticsHost = {
    getCurrentDirectory: () => tsc.sys.getCurrentDirectory(),
    getCanonicalFileName: fileName => fileName,
    getNewLine: () => tsc.sys.newLine,
  };

  return tsc.formatDiagnostics(errors, defaultFormatHost);
}

function readCompilerOptions(
  configPath: string,
  rootDir: string,
): tsc.CompilerOptions {
  configPath = path.resolve(rootDir, configPath);
  const { config, error } = tsc.readConfigFile(configPath, tsc.sys.readFile);
  if (error) {
    throw error;
  }

  const { errors, options } = tsc.parseJsonConfigFileContent(
    config,
    tsc.sys,
    path.resolve(rootDir),
  );

  if (errors.length > 0) {
    const formattedErrors = formatTsDiagnostics(errors);
    throw new Error(
      `Some errors occurred while attempting to read from ${configPath}: ${formattedErrors}`,
    );
  }

  return options;
}

function getStartDir(): string {
  // This is needed because of the way our tests are structured.
  // If this is being executed as a library (under node_modules)
  // we want to start with the project directory that's three
  // levels above.
  // If this is being executed from the test suite, we want to start
  // in the directory of the test

  const grandparent = path.resolve(__dirname, '..', '..');
  if (grandparent.endsWith(`${path.sep}node_modules`)) {
    return process.cwd();
  }

  return '.';
}

function getPathToClosestTSConfig(
  startDir?: string,
  previousDir?: string,
): string {
  // Starting with the startDir directory and moving on to the
  // parent directory recursively (going no further than the root directory)
  // find and return the path to the first encountered tsconfig.json file

  if (!startDir) {
    return getPathToClosestTSConfig(getStartDir());
  }

  const tsConfigPath = path.join(startDir, 'tsconfig.json');

  const startDirPath = path.resolve(startDir);
  const previousDirPath = path.resolve(previousDir || '/');

  if (startDirPath === previousDirPath || fs.existsSync(tsConfigPath)) {
    return tsConfigPath;
  }

  return getPathToClosestTSConfig(path.join(startDir, '..'), startDir);
}

function getTSConfigPathFromConfig(globals: ConfigGlobals): string {
  const defaultTSConfigFile = getPathToClosestTSConfig();
  if (!globals) {
    return defaultTSConfigFile;
  }

  const tsJestConfig = getTSJestConfig(globals);

  if (tsJestConfig.tsConfigFile) {
    return tsJestConfig.tsConfigFile;
  }

  return defaultTSConfigFile;
}

export function mockGlobalTSConfigSchema(
  globals: ConfigGlobals,
): ConfigGlobals {
  const configPath = getTSConfigPathFromConfig(globals);
  return { 'ts-jest': { tsConfigFile: configPath } };
}

export const getTSConfig = _.memoize(getTSConfig_local, (globals, rootDir) => {
  // check cache before resolving configuration
  // NB: We use JSON.stringify() to create a consistent, unique signature. Although it lacks a uniform
  //     shape, this is simpler and faster than using the crypto package to generate a hash signature.
  return JSON.stringify(globals, rootDir);
});

// Non-memoized version of TSConfig
function getTSConfig_local(globals, rootDir: string = '') {
  const configPath = getTSConfigPathFromConfig(globals);
  logOnce(`Reading tsconfig file from path ${configPath}`);
  const skipBabel = getTSJestConfig(globals).skipBabel;

  const config = readCompilerOptions(configPath, rootDir);
  logOnce('Original typescript config before modifications: ', { ...config });

  // ts-jest will map lines numbers properly if inlineSourceMap and
  // inlineSources are set to true. The sourceMap configuration
  // is used to send the sourcemap back to Jest
  delete config.sourceMap;
  config.inlineSourceMap = true;
  config.inlineSources = true;

  // the coverage report is broken if `.outDir` is set
  // see https://github.com/kulshekhar/ts-jest/issues/201
  // `.outDir` is removed even for test files as it affects with breakpoints
  // see https://github.com/kulshekhar/ts-jest/issues/309
  delete config.outDir;

  if (configPath === path.join(getStartDir(), 'tsconfig.json')) {
    // hardcode module to 'commonjs' in case the config is being loaded
    // from the default tsconfig file. This is to ensure that coverage
    // works well. If there's a need to override, it can be done using
    // a custom tsconfig for testing
    config.module = tsc.ModuleKind.CommonJS;
  }

  config.module = config.module || tsc.ModuleKind.CommonJS;
  config.jsx = config.jsx || tsc.JsxEmit.React;

  if (config.allowSyntheticDefaultImports && !skipBabel) {
    // compile ts to es2015 and transform with babel afterwards
    config.module = tsc.ModuleKind.ES2015;
  }

  return config;
}

export function cacheFile(
  jestConfig: JestConfig,
  filePath: string,
  src: string,
): void {
  // store transpiled code contains source map into cache, except test cases
  if (!jestConfig.testRegex || !filePath.match(jestConfig.testRegex)) {
    const outputFilePath = path.join(
      jestConfig.cacheDirectory,
      '/ts-jest/',
      crypto
        .createHash('md5')
        .update(filePath)
        .digest('hex'),
    );

    fsExtra.outputFileSync(outputFilePath, src);
  }
}

export function runTsDiagnostics(
  filePath: string,
  compilerOptions: tsc.CompilerOptions,
): void {
  const program = tsc.createProgram([filePath], compilerOptions);
  const allDiagnostics = tsc.getPreEmitDiagnostics(program);

  if (allDiagnostics.length) {
    throw new Error(formatTsDiagnostics(allDiagnostics));
  }
}
