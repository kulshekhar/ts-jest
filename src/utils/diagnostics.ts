export const TsJestDiagnosticCodes = {
  Generic: 151000,
  ConfigModuleOption: 151001,
  ModernNodeModule: 151002,
} as const
