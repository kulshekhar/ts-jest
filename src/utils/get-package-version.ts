/**
 * @internal
 */
export function getPackageVersion(moduleName: string): string | undefined {
  try {
    return require(`${moduleName}/package.json`).version as string
  } catch (err) {}

  return
}
