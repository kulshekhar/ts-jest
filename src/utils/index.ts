export { mocked } from './testing'
export { createJestPreset } from '../presets/create-jest-preset'
export { pathsToModuleNameMapper } from '../config/paths-to-module-name-mapper'
