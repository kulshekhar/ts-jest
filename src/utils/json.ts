/* eslint-disable no-redeclare */
import stableStringify = require('fast-json-stable-stringify')

const UNDEFINED = 'undefined'

export function stringify(input: unknown): string {
  return input === undefined ? UNDEFINED : stableStringify(input)
}

export function parse(input: string): any {
  return input === UNDEFINED ? undefined : JSON.parse(input)
}

interface NormalizeOptions {
  parse?: (input: string) => any
}
/**
 * @internal
 */
export function normalize(input: string, { parse: parser = parse }: NormalizeOptions = {}): string {
  let result: string | undefined
  if (normalize.cache.has(input)) {
    result = normalize.cache.get(input)
  } else {
    const data = parser(input)
    result = stringify(data)
    if (result === input) result = undefined
    normalize.cache.set(input, result)
  }

  return result === undefined ? input : result
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace normalize {
  export const cache = new Map<string, string | undefined>()
}
