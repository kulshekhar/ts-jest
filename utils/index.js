const { pathsToModuleNameMapper } = require('../dist/config/paths-to-module-name-mapper')
const { createJestPreset } = require('../dist/presets/create-jest-preset')
const { mocked } = require('../dist/utils/testing')

module.exports = {
  get mocked() {
    if (!process.env.DISABLE_MOCKED_WARNING) {
      console.warn(
        '\n`mocked` util function is now deprecated and has been moved to Jest repository,' +
          ' see https://github.com/facebook/jest/pull/12089. In `ts-jest` v28.0.0, `mocked` function will be completely removed.' +
          ' Users are encouraged to use to Jest v27.4.0 or above to have `mocked` function available from `jest-mock`. ' +
          'One can disable this warning by setting environment variable process.env.DISABLE_MOCKED_WARNING=true\n'
      )
    }

    return mocked
  },
  get createJestPreset() {
    console.warn('ts-jest[main] (WARN) Replace any occurrences of "ts-jest/utils" with just "ts-jest".')

    return createJestPreset
  },
  get pathsToModuleNameMapper() {
    console.warn('ts-jest[main] (WARN) Replace any occurrences of "ts-jest/utils" with just "ts-jest".')

    return pathsToModuleNameMapper
  },
}
