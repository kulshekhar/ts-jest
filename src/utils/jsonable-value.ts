import { stringify } from './json'

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class JsonableValue<V = Record<string, any>> {
  private _serialized!: string
  private _value!: V

  constructor(value: V) {
    this.value = value
  }

  set value(value: V) {
    this._value = value
    this._serialized = stringify(value)
  }

  get value(): V {
    return this._value
  }

  get serialized(): string {
    return this._serialized
  }

  valueOf(): V {
    return this._value
  }

  toString(): string {
    return this._serialized
  }
}
