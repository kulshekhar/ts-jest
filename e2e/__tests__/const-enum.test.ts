import runJest, { json as runWithJson } from '../run-jest'
import { extractSortedSummary } from '../utils'

const DIR_NAME = 'const-enum'

test(`successfully runs the tests inside ${DIR_NAME} with 'isolatedModules: false'`, () => {
  const { json } = runWithJson(DIR_NAME)

  expect(json.success).toBe(true)
})

test(`partial successfully runs the tests inside ${DIR_NAME} with 'isolatedModules: true'`, () => {
  const result = runJest(DIR_NAME, ['-c=jest-isolated.config.js'])

  expect(extractSortedSummary(result.stderr).rest).toMatchInlineSnapshot(`
    "FAIL __tests__/import-from-d-ts-no-js.spec.ts
      ● Test suite failed to run

        Cannot find module '../hoo-constant' from '__tests__/import-from-d-ts-no-js.spec.ts'

        However, Jest was able to find:
        	'../hoo-constant.d.ts'

        You might want to include a file extension in your import, or update your 'moduleFileExtensions', which is currently ['js', 'jsx', 'ts', 'tsx', 'json', 'node'].

        See https://jestjs.io/docs/configuration#modulefileextensions-arraystring

        > 1 | import { HooConstEnum } from '../hoo-constant'
            | ^
          2 |
          3 | const getTwo = (): string => HooConstEnum.two
          4 |

          at Resolver.resolveModule (../../node_modules/jest-resolve/build/resolver.js:322:11)
          at Object.<anonymous> (__tests__/import-from-d-ts-no-js.spec.ts:1:1)

    PASS __tests__/import-from-d-ts-has-js.spec.ts
    PASS __tests__/import-from-ts.spec.ts"
  `)
})
