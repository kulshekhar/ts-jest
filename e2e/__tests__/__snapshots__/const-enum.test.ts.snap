// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`partial successfully runs the tests inside const-enum with 'isolatedModules: true' 1`] = `
"PASS __tests__/import-from-d-ts-has-js.spec.ts
PASS __tests__/import-from-d-ts-no-js.spec.ts
PASS __tests__/import-from-ts.spec.ts"
`;
